Identify the theme/plugin responsible for calling amp_is_available() too early

westonruter opened the pull request:
## Summary

Fixes #4602

Let's say that Akismet called `is_amp_endpoint()` too early.

Before, the usage notice would not even mention Akismet:

> PHP Notice:  amp_is_available was called <strong>incorrectly</strong>. amp_is_available() (or amp_is_request(), formerly is_amp_endpoint()) was called too early and so it will not work properly. WordPress is currently doing the "init" action. Calling this function before the "wp" action means it will not have access to WP_Query and the queried object to determine if it is an AMP response, thus neither the "amp_skip_post()" filter nor the AMP enabled toggle will be considered. Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 2.0.0.) in /app/public/core-dev/src/wp-includes/functions.php on line 5225

After, this is the message where it explicitly mentions Akismet:

> PHP Notice:  amp_is_available was called <strong>incorrectly</strong>. `amp_is_available()` (or `amp_is_request()`, formerly `is_amp_endpoint()`) was called too early and so it will not work properly. It appears the `plugin` with slug `akismet` is responsible; please contact the author. WordPress is currently doing the `init` hook. Calling this function before the `wp` action means it will not have access to `WP_Query` and the queried object to determine if it is an AMP response, thus neither the `amp_skip_post()` filter nor the AMP enabled toggle will be considered. Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 2.0.0.) in /app/public/core-dev/src/wp-includes/functions.php on line 5225

Similarly, if Twenty Twenty called `is_amp_endpoint()` too early:

> PHP Notice:  amp_is_available was called <strong>incorrectly</strong>. `amp_is_available()` (or `amp_is_request()`, formerly `is_amp_endpoint()`) was called too early and so it will not work properly. It appears the `theme` with slug `twentytwenty` is responsible; please contact the author. WordPress is currently doing the `after_setup_theme` hook. Calling this function before the `wp` action means it will not have access to `WP_Query` and the queried object to determine if it is an AMP response, thus neither the `amp_skip_post()` filter nor the AMP enabled toggle will be considered. Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 2.0.0.) in /app/public/core-dev/src/wp-includes/functions.php on line 5225

**Update (by @schlessera)**

This PR now also does the following:
- Refactors services to provide a subnamespace `DevTools`.
- Adds `FileReflection` and `CallbackReflection` helper objects to get source data.
- Adds `LikelyCulpritDetector` to step through a trace stack and return the most likely culprit of that stack.
- Uses these helper objects in `is_amp_endpoint()` to clean up the detection logic.
- Uses these helper objects in `ErrorPage` to provide information about the most likely culprit as well.

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

westonruter committed:
Identify the theme/plugin responsible for calling amp_is_available() too early

westonruter commented on includes/validation/class-amp-validation-manager.php:
```diff
@@ -1585,11 +1586,53 @@ public static function get_source( $callback ) {
+	 * }
+	 */
+	public static function get_file_source( $file ) {
+		if (
+			! isset( self::$template_directory )
+			||
+			! isset( self::$template_slug )
+			||
+			! isset( self::$stylesheet_directory )
+			||
+			! isset( self::$stylesheet_slug )
+		) {
+			self::set_theme_variables();
+		}
```
I don't like this.

schlessera replied:
What about moving the checks within the `set_theme_variables()` method?

schlessera replied:
The ugly part here is that `set_theme_variables() `is an initialization method (like a constructor) that needs to be called for each execution of `get_file_source()` because the class is static and doesn't have any instantiation with a constructor.

To get around this, you could turn these static members into static methods with a caching static variable instead:
So, instead of referring to `self::$stylesheet_directory` in other parts of the code, you'd refer to `self::get_stylesheet_directory()` instead. Within these static methods, you can then do the initialization once and store it in a static variable.

However, I think that you're just trading multiple checks to a function call, so it's probably not a performance gain, only cleaner code.

Refactoring the entire validation manager into an instanced object (or a series of them) would be the solution that also improves performance.

westonruter replied:
Something else that's tricky is that we need to wait until as late as possible to determine these values.

During validation requests, note how `set_theme_variables` gets called twice overall:

https://github.com/ampproject/amp-wp/blob/4ee9ee4ecc48431a26690245123b73f4b7154d1f/includes/validation/class-amp-validation-manager.php#L566-L570

This ensures that the theme is accurate when it is dynamically switched by plugins.

And calling it again when emitting the warning here will make sure that the theme variables get set to reflect the current state at the time the error happened.

schlessera replied:
As it seems to make sense to reuse this in `ErrorPage`, I'd suggest we extract this out into a helper object instead.

westonruter committed:
Wrap code refs in _doing_it_wrong in backticks

westonruter commented:
I'm punting this after 2.0.1 since it's not directly fixing a bug. That being said, if reviewed and approved we can put it back in 2.0.2.

schlessera commented on includes/amp-helper-functions.php:
```diff
@@ -411,23 +411,60 @@ function amp_is_available() {
+		if ( ! empty( $closest_source['type'] ) && ! empty( $closest_source['name'] ) ) {
+			$message .= ' ' . sprintf(
+				/* translators: 1: type (theme, plugin, must-use plugin), 2: name */
+				__( 'It appears the %1$s with slug %2$s is responsible; please contact the author.', 'amp' ),
+				'`' . $closest_source['type'] . '`',
+				'`' . $closest_source['name'] . '`'
```
You're using a non-translatable value (`type`) as a word that is part of the sentence to translate, so in any other language, the English "type" will appear.

You'd instead need a separate translatable sentence per possible type to consider here.

westonruter replied:
Yes, I thought about that. But I decided to not go to that effort since this is an error message. The type will correspond to directory names that a WP site admin will already be familiar with, like `plugin`, `theme`, and `mu-plugin`. For that reason, I thought translating these would be overkill.

westonruter replied:
In other words, WordPress doesn't localize its filesystem.

schlessera replied:
A regular user creating content on their site might never have seen the WordPress filesystem. All they see is the translated menu entries and UI elements.

westonruter replied:
Where would regular users see the `_doing_it_wrong()` notices? Wouldn't it be in the context of PHP errors? There isn't otherwise a UI for displaying this to regular users (yet).

I'm not against translating these. I was just cutting a corner which didn't seem super important.

schlessera replied:
Right, they wouldn't see this for the `doing_it_wrong()` part. But I reused this for the `ErrorPage` which is shown if AMP processing failed. This can indeed be shown to regular users.

I can update the output in the `ErrorPage` independently.

westonruter replied:
Oh yes, of course 👍 

One thing to keep keep in mind for translation is that there will probably need to be three translation strings:

* `It appears the plugin with slug %s is responsible.`
* `It appears the must-use plugin with slug %s is responsible.`
* `It appears the theme with slug %s is responsible.`

As I know you're intimately aware, the “the” will translate differently depending on what it is defining.

I think it's useful to go include the slug for sure, but also we could include the name. For development purposes, the slug is much more helpful since it makes it easy to know where to `cd`. For non-development purposes, the name is more helpful to identify something to deactivate.

schlessera replied:
I adapted this for `ErrorPage` but then went ahead and modified the text in `is_amp_endpoint()` as well. Even though the translation was technically fine for `is_amp_endpoint()` in the expected developer context, it makes sense to have shared translations across these two.

schlessera replied:
I didn't put the plugin/theme name into the string just yet and kept the slug for now. The reason is that the field in the `source` array that is used for storing the slug is called `name`. Adding the actual name in addition to the slug would therefore require a refactor that touches a lot of code, so this is best done in a separate PR. This one is messy enough as it is.

schlessera commented on includes/amp-helper-functions.php:
```diff
@@ -411,23 +411,60 @@ function amp_is_available() {
+		);
+
+		if ( ! empty( $closest_source['type'] ) && ! empty( $closest_source['name'] ) ) {
+			$message .= ' ' . sprintf(
```
I think that this way of assembling the text will lead to problems with RTL languages, as you're encoding the order of the sentences with forced spaces.

westonruter replied:
I don't think so. RTL languages are concatenated with spaces as well. A space is agnostic to the text direction.

schlessera commented on includes/validation/class-amp-validation-manager.php:
```diff
@@ -1585,11 +1586,52 @@ public static function get_source( $callback ) {
+	 * }
+	 */
+	public static function get_file_source( $file ) {
+		if (
+			! isset( self::$template_directory )
+			||
+			! isset( self::$template_slug )
+			||
+			! isset( self::$stylesheet_directory )
+			||
+			! isset( self::$stylesheet_slug )
+		) {
```
These conditions can be combined:
```suggestion
		if ( ! isset( self::$template_directory, self::$template_slug, self::$stylesheet_directory, self::$stylesheet_slug ) ) {
```

westonruter replied:
Not quite, as that combined usage would be the same as:

```php
			! isset( self::$template_directory )
			&&
			! isset( self::$template_slug )
			&&
			! isset( self::$stylesheet_directory )
			&&
			! isset( self::$stylesheet_slug )
```

westonruter replied:
Alternatively, we could eliminate this condition and invocation of `set_theme_variables()` here entirely, and instead call `AMP_Validation_Manager::set_theme_variables()` from inside of `amp_is_available()`. That seems somewhat better. Or `get_file_source()` could take a second `$update_theme_variables` arg which when set would call the method.

schlessera replied:
From the [`isset` docs](https://www.php.net/manual/en/function.isset.php):

> If multiple parameters are supplied then isset() will return TRUE only if all of the parameters are considered set. Evaluation goes from left to right and stops as soon as an unset variable is encountered.

So it will indeed be the same as what you had.

schlessera replied:
Maybe this is confusing because of the negation. Do note that in the combined `isset()` the entire construct is negated as a whole, not every individual check.

So, `isset( a, b, c )` will be **true** if all of a, b & c are set => will be **false** if **either a or b or c is not set**.
Therefore `! isset( a, b, c )` will be **false** if all of a, b & c are set => will be **true** if **either a or b or c is not set**.

schlessera replied:
Or to paraphrase in the code you've used as the bad example above...
It's not this:
```php
			! isset( self::$template_directory )
			&&
			! isset( self::$template_slug )
			&&
			! isset( self::$stylesheet_directory )
			&&
			! isset( self::$stylesheet_slug )
```
but rather this:
```php
			! (
				isset( self::$template_directory )
				&&
				isset( self::$template_slug )
				&&
				isset( self::$stylesheet_directory )
				&&
				isset( self::$stylesheet_slug )
			)
```
which is equivalent to your original code.

westonruter replied:
Yes, it confuses me. 

I've simplified it by removing the conditionals entirely in 64c84073a.

schlessera commented on includes/amp-helper-functions.php:
```diff
@@ -411,23 +411,60 @@ function amp_is_available() {
+			);
+		}
+
+		$message .= ' ' . sprintf(
+			/* translators: 1: the current action, 2: the wp action, 4: the WP_Query class, 4: the amp_skip_post() function */
+			__( 'WordPress is currently doing the %1$s action. Calling this function before the %2$s action means it will not have access to %3$s and the queried object to determine if it is an AMP response, thus neither the %4$s filter nor the AMP enabled toggle will be considered.', 'amp' ),
+			'`' . current_action() . '`',
+			'`wp`',
+			'`WP_Query`',
+			'`amp_skip_post()`'
 		);
```
I think this part of the text should come before the `It appears the %1$s with slug %2$s is responsible; please contact the author.` part. That would read better. Right now we have:
1. timing problem
2. root cause
3. timing reason

If you switch 2. and 3., it will be easier to follow.

westonruter replied:
As in d047556?

westonruter committed:
Improve order of message construction

westonruter commented:
Rebasing against `develop`…

schlessera committed:
Create new sub-namespace for DevTools

westonruter committed:
Simplify ensuring that theme variables are set

schlessera commented:
@googlebot I consent.

schlessera committed:
Fix circular reference detection in injector

schlessera committed:
Add FileReflection and CallbackReflection

schlessera committed:
Update PHPStorm meta file

schlessera committed:
Update code to new structure

schlessera committed:
Update tests

schlessera committed:
Fix PHPCS issues

schlessera committed:
Make PHPStan understand the Services container

schlessera committed:
Fix remaining PHPStan issues

schlessera commented on includes/amp-helper-functions.php:
```diff
@@ -414,23 +414,60 @@ function amp_is_available() {
+			'name' => '',
+		];
+
+		foreach ( debug_backtrace( 0 ) as $call_stack ) { // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_debug_backtrace
```
@westonruter You're passing `0` here to the `$options` argument of `debug_backtrace()`. Just to be clear, I assume this was meant to lift the second argument `$limit` ?

schlessera replied:
...or was the goal to unset the default `DEBUG_BACKTRACE_PROVIDE_OBJECT` option to save processing time? I think that would actually be the wanted behavior here as well.
Using `0` here is very ambiguous, though, and will warrant at least a clarifying comment.

I'll adapt as needed.

schlessera committed:
Show the likely culprit on the ErrorPage

schlessera committed:
Code quality fixes

schlessera committed:
Make tests for plugin suppression independent of other tests

schlessera committed:
Change plugin registry test setup

westonruter commented on includes/admin/class-amp-template-customizer.php:
```diff
@@ -75,7 +75,6 @@ protected function __construct( WP_Customize_Manager $wp_customize, ReaderThemeL
 	 * @return AMP_Template_Customizer Instance.
 	 */
 	public static function init( WP_Customize_Manager $wp_customize ) {
-		/** @var ReaderThemeLoader $reader_theme_loader */
```
Note that this would still be useful for non-IntelliJ based editors.

schlessera replied:
Do you think we should keep them? I think we're all using IntelliJ on the team, so that would be for external contributors mostly.

westonruter replied:
I don't really care. I'm fine with removing them. Just wanted to point out there is a bit of utility to having them.

westonruter committed:
Fix tests by accounting for varying plugin_dir

westonruter committed:
Opt instead to nullify plugin_file_pattern

schlessera committed:
Exclude PHPStan extensions from compat checks

schlessera committed:
Add path exclude to PHP 5.6 linting

schlessera committed:
Revert data backup change in PluginRegistry test

schlessera committed:
Improve translations for culprit message on ErrorPage

schlessera committed:
Adapt cuplrit test in is_amp_endpoint() to use same translations as ErrorPage

schlessera commented:
@weston I think this is ready for another review (can't assign to you as you're the author).

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -413,24 +414,59 @@ function amp_is_available() {
 		);
+
+		$message .= ' ' . sprintf(
+			/* translators: 1: the current action, 2: the wp action, 4: the WP_Query class, 4: the amp_skip_post() function */
+			__( 'WordPress is currently doing the %1$s action. Calling this function before the %2$s action means it will not have access to %3$s and the queried object to determine if it is an AMP response, thus neither the %4$s filter nor the AMP enabled toggle will be considered.', 'amp' ),
```
`current_action()`  is an alias for `current_filter()` so this could be an action or a filter.

```suggestion
			/* translators: 1: the current hook, 2: the wp action, 4: the WP_Query class, 4: the amp_skip_post() function */
			__( 'WordPress is currently doing the %1$s hook. Calling this function before the %2$s action means it will not have access to %3$s and the queried object to determine if it is an AMP response, thus neither the %4$s filter nor the AMP enabled toggle will be considered.', 'amp' ),
```

westonruter committed:
Reference hooks instead of action since could be filter

westonruter commented:
Rebasing onto `develop`…

westonruter commented:
@schlessera I assume you ✅  now that you've made additional changes?

@pierlon I'd appreciate a review from you as well.

westonruter commented:
Re-rebasing against `develop`.

westonruter commented:
I tried adding `is_amp_endpoint()` to `twentytwenty_theme_support()` and I successfully got:

> PHP Notice:  amp_is_available was called <strong>incorrectly</strong>. `amp_is_available()` (or `amp_is_request()`, formerly `is_amp_endpoint()`) was called too early and so it will not work properly. WordPress is currently doing the `after_setup_theme` hook. Calling this function before the `wp` action means it will not have access to `WP_Query` and the queried object to determine if it is an AMP response, thus neither the `amp_skip_post()` filter nor the AMP enabled toggle will be considered. It appears the theme with slug `twentytwenty` is responsible; please contact the author. Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 2.0.0.) in /app/public/core-dev/src/wp-includes/functions.php on line 5225

I also added `is_amp_endpoint()` to `Jetpack::deprecated_hooks()` and I successfully got:

> PHP Notice:  amp_is_available was called <strong>incorrectly</strong>. `amp_is_available()` (or `amp_is_request()`, formerly `is_amp_endpoint()`) was called too early and so it will not work properly. WordPress is currently doing the `init` hook. Calling this function before the `wp` action means it will not have access to `WP_Query` and the queried object to determine if it is an AMP response, thus neither the `amp_skip_post()` filter nor the AMP enabled toggle will be considered. It appears the plugin with slug `jetpack` is responsible; please contact the author. Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 2.0.0.) in /app/public/core-dev/src/wp-includes/functions.php on line 5225

I also added a `foo.php` to the mu-plugins which calls `is_amp_endpoint()` at `init` and I got:

> PHP Notice:  amp_is_available was called <strong>incorrectly</strong>. `amp_is_available()` (or `amp_is_request()`, formerly `is_amp_endpoint()`) was called too early and so it will not work properly. WordPress is currently doing the `init` hook. Calling this function before the `wp` action means it will not have access to `WP_Query` and the queried object to determine if it is an AMP response, thus neither the `amp_skip_post()` filter nor the AMP enabled toggle will be considered. It appears the must-use plugin with slug `foo.php` is responsible; please contact the author. Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 2.0.0.) in /app/public/core-dev/src/wp-includes/functions.php on line 5225

westonruter committed:
Check theme sources before plugins since plugin may register theme directory inside itself

westonruter committed:
Deprecate AMP_Validation_Manager::get_source()

westonruter committed:
De-duplicate ReflectionMethod instantiation

westonruter committed:
Add tests for CallbackReflection

westonruter committed:
Test get_source() for function that does not exist

westonruter committed:
Remove bogus PHPUnit\Framework\string type hint caused by setting IDE to PHP 5.6

westonruter committed:
Add test for getting source of child theme function

westonruter committed:
Add test theme functions

westonruter committed:
Add tests for FileReflection

westonruter committed:
Make use of injector service during test setUp

westonruter committed:
Skip using injector since FileReflection needs a register() call

schlessera commented on tests/php/src/DevTools/CallbackReflectionTest.php:
```diff
@@ -39,7 +39,10 @@ public function setUp() {
-		$this->callback_reflection = Services::get( 'injector' )->make( CallbackReflection::class );
+		$plugin_registry = new PluginRegistry();
+		$file_reflection = new FileReflection( $plugin_registry );
+		$file_reflection->register();
```
Not necessarily something you should do here, but wanted to clarify nevertheless:

The `FileReflection` service is a "shared" instance: https://github.com/ampproject/amp-wp/pull/5289/commits/886022d51bbcfe463c507e33aa8981e81e5c674d#diff-e12f38193bbbb46cce767c749eaebf1bR146-R153

This means that both the injector and the service locator will always return the same object instance. So you could for example use the injector to make `CallbackReflection` and then use the service locator to call the `register()` method on the `FileReflection` and it would hit the one that was already injected into the `CallbackReflection`.

schlessera replied:
No idea where this link is actually going to... 😕 

Here's what I wanted to show:
![Image 2020-09-13 at 10 16 12 AM](https://user-images.githubusercontent.com/83631/93013615-345d5a00-f5aa-11ea-953d-4aba7dfd0bd3.png)

westonruter committed:
Reset parent_theme_pattern and child_theme_pattern; test get_file_source() for unknown file

westonruter committed:
Add test stubs for LikelyCulpritDetector

westonruter committed:
Fix PhpStan analysis for resetting child_theme_pattern and parent_theme_pattern

westonruter committed:
Add missing Services use statement

schlessera committed:
Add injector to PHPStorm meta file

schlessera committed:
Include injector in list of services that PHPStan knows about

schlessera committed:
Add tests for analyze_trace() and analyze_exception()

schlessera committed:
Deduplicate file stack to trace conversion

schlessera committed:
Fix alignment

schlessera committed:
Add test for analyze_backtrace()

westonruter commented on src/DevTools/LikelyCulpritDetector.php:
```diff
@@ -0,0 +1,95 @@
+				||
+				'core' === $source['type']
+				||
+				( 'plugin' === $source['type'] && 'amp' === $source['name'] )
```
Just realized we should include Gutenberg here.

westonruter replied:
Nevertheless, currently Gutenberg is highly unlikely to be involved at present for such caught errors. For validation errors, Gutenberg is omitted from being a culprit via:

https://github.com/ampproject/amp-wp/blob/0ea1d1635182462d8765e1a32448f1a0251a52f2/includes/validation/class-amp-validated-url-post-type.php#L1391-L1394

But if `LikelyCulpritDetector` is extended to be used for AMP validation purposes (as it probably should) then this will be accounted for if we exempt Gutenberg here as well.

westonruter commented on src/DevTools/LikelyCulpritDetector.php:
```diff
@@ -0,0 +1,95 @@
+	 */
+	public function analyze_exception( Exception $exception ) {
+		$trace = $exception->getTrace();
+		array_unshift( $trace, [ 'file' => $exception->getFile() ] );
```
This is needed because `getTrace()` doesn't include the call stack that raised the exception?

westonruter replied:
[Yes](https://www.php.net/manual/en/exception.gettrace.php#Hcom107563:~:text=%20The%20trace%20does%20not%20include,recorded%20in%20the%20top%2Dlevel%20getFile%2FLine%20methods.): 

> 1) The trace does not include the file / line at which the exception is thrown; that entry is only recorded in the top-level getFile/Line methods.

westonruter commented:
Error page changes are working well. ✅ 

# Must-Use Plugin

```php
<?php
/**
 * Plugin Name: my-mu-plugin-schemaorg-filter-exception-thrower
 */

add_filter(
	'amp_schemaorg_metadata',
	function () {
		throw new Exception( 'I am a bad thrower! I am a mu-plugin!' );
	}
);
```

![image](https://user-images.githubusercontent.com/134745/93025971-45f92e80-f5b7-11ea-8002-c76a80fbcf21.png)

# Plugin

```php
<?php
/**
 * Plugin Name: my-plugin-schemaorg-filter-exception-thrower
 */

add_filter(
	'amp_schemaorg_metadata',
	function () {
		throw new Exception( 'I am a bad thrower! I am a plugin!' );
	}
);
```

![image](https://user-images.githubusercontent.com/134745/93026005-8bb5f700-f5b7-11ea-8f2f-327ce45e1b4e.png)

# Theme

Adding this to Twenty Twenty's `functions.php`:

```php
add_filter(
	'amp_schemaorg_metadata',
	function () {
		throw new Exception( 'I am a bad thrower! I am a theme!' );
	}
);
```

![image](https://user-images.githubusercontent.com/134745/93026017-ab4d1f80-f5b7-11ea-911e-46696dfcb0ca.png)

westonruter commented on tests/php/src/DevTools/LikelyCulpritDetectorTest.php:
```diff
@@ -0,0 +1,283 @@
+						get_stylesheet_directory() . '/functions.php',
+					],
+					'mu-plugin',
+					// TODO: Is this a correct slug for a single file (MU) plugin?
```
Yes it is. It's like a non-mu plugin which is not in a directory.

westonruter commented on tests/php/src/DevTools/LikelyCulpritDetectorTest.php:
```diff
@@ -0,0 +1,283 @@
+			'mu-plugin'    => [
+				[ WP_CONTENT_DIR . '/mu-plugins/bad-mu-plugin.php' ],
+				'mu-plugin',
+				// TODO: Is this a correct slug for a single file (MU) plugin?
```
Yes.

westonruter committed:
Consider Gutenberg as core like AMP; use constants and remove todos

westonruter commented:
Checking AMP validation. ✅ 

# Must Use Plugin

```php
<?php
/**
 * Plugin Name: my-mu-plugin-bad-footer
 */

add_action( 'wp_footer', function () {
	echo "<mu-plugin></mu-plugin>";
} );
```

![image](https://user-images.githubusercontent.com/134745/93026557-a2f6e380-f5bb-11ea-85e7-7dd4afa7ac01.png)

# Plugin

```php
<?php
/**
 * Plugin Name: my-mu-plugin-bad-footer
 */

add_action( 'wp_footer', function () {
	echo "<plugin></plugin>";
} );
```

![image](https://user-images.githubusercontent.com/134745/93026578-cde13780-f5bb-11ea-8dd7-d5e3ba09e016.png)

# Theme

Adding this code to Twenty Twenty's `functions.php`:

```php
add_action( 'wp_footer', function () {
	echo "<theme></theme>";
} );
```

![image](https://user-images.githubusercontent.com/134745/93026601-fb2de580-f5bb-11ea-8980-3af50e37546b.png)

westonruter commented:
Related issue I found when testing: https://github.com/ampproject/amp-wp/pull/5375

schlessera commented on src/DevTools/LikelyCulpritDetector.php:
```diff
@@ -54,7 +54,7 @@ public function analyze_backtrace() {
 	public function analyze_exception( Exception $exception ) {
 		$trace = $exception->getTrace();
-		array_unshift( $trace, [ 'file' => $exception->getFile() ] );
+		array_unshift( $trace, [ FileReflection::SOURCE_FILE => $exception->getFile() ] );
```
I was hesitant about these constants. I think the `source` array (which is reused in many different places) should at one point become a `Source` object instead.

These constants here won't hurt for now, and this is all marked as @internal anyway. But the long-term approach should be to refactor all the `source` array shenanigans into an object, so that these constants won't be needed as we'll have methods instead.

schlessera replied:
What we could do as an intermediate step is to extract an interface `Source` that receives these constants. They would then look like `Source::TYPE`, so a bit less unwieldy.

Later on, we can then change this `Source` interface into a `Source` class to turn the arrays into objects (or even keep it as an interface and have differing implementations if that makes sense).

schlessera replied:
The main benefit of that intermediate step is that we not only move the hard-coded strings into constants and thereby have a "catalogue of options", we also do so in such a way that it can be kept with the move to objects. No matter what `Source` evolves into, the constants can stay within it.

westonruter commented on src/DevTools/ErrorPage.php:
```diff
@@ -235,6 +253,41 @@ private function render_message() {
+					return '';
+			}
+
+			return sprintf(
```
IntelliJ is giving some static analysis warning here:

![image](https://user-images.githubusercontent.com/134745/93028049-169dee00-f5c6-11ea-97fd-8f82b42435b1.png)

westonruter replied:
Addressed in 8369305ea1ddc58d20c44eca00c85ea1b8f42adf.

westonruter commented on src/DevTools/ErrorPage.php:
```diff
@@ -262,10 +315,12 @@ private function render_exception() {
 			get_class( $this->exception ),
+			$this->exception->getFile(),
+			$this->exception->getLine(),
 			str_replace( "\n", '<br>', $this->exception->getTraceAsString() )
```
No need for `<br>` in a `<pre>`

westonruter replied:
Done in 8369305ea1ddc58d20c44eca00c85ea1b8f42adf

westonruter committed:
Improve error page output generation for static analysis, escaping, newlines

westonruter committed:
Remove erroneous commenting-out of Gutenberg

westonruter committed:
Use elseif instead in get_file_source

schlessera commented on src/DevTools/FileReflection.php:
```diff
@@ -176,41 +176,33 @@ public function get_file_source( $file ) {
-		}
-
-		if ( $this->is_child_theme_file( $file, $matches ) ) {
+		} elseif ( $this->is_child_theme_file( $file, $matches ) ) {
```
Why this change to `elseif`s? The individual conditions already end in `return`s...

westonruter replied:
Reverted as discussed. As you shared: https://github.com/kalessil/phpinspectionsea/blob/master/docs/control-flow.md#redundant-else-keyword

westonruter committed:
Flesh out array return types

westonruter committed:
Revert "Use elseif instead in get_file_source"

This reverts commit ca7c31126dabaaff60d9eb0e48fc24dc58353a75.

westonruter commented on tests/php/src/DevTools/LikelyCulpritDetectorTest.php:
```diff
@@ -0,0 +1,281 @@
+	private function get_trace_from_file_stack( $file_stack ) {
+		return array_map(
+			static function ( $file ) {
+				return [ 'file' => $file ];
```
This could also be:

```php
return compact( 'file' );
```

I know `extract()` is bad. Any reason to not use `compact()`?

schlessera replied:
I opted to go with a simple array for performance reasons. The gain in simplicity is not worth the performance impact for a single variable. `compact()` does a lot of processing, first of all because you pass in a string name instead of using the variable directly. This needs to be parsed and mapped first to be able to assign.

If you have multiple variables (or a dynamic set) though, `compact()` is certainly fine and doesn't have the security implications of `extract()`.
